r/R/filesystem.R

```diff
@@ -233,7 +233,11 @@ FileSystem <- R6Class("FileSystem", inherit = Object,
 #' @export
 LocalFileSystem <- R6Class("LocalFileSystem", inherit = FileSystem)
 LocalFileSystem$create <- function() {
-  shared_ptr(LocalFileSystem, fs___LocalFileSystem__create())
+  out <- shared_ptr(LocalFileSystem, fs___LocalFileSystem__create())
+  # HACK: For some reason, the filesystem APIs use raw pointers, not shared_ptr
+  # so we have to preserve them in some more global scope to use them as expected
+  options(arrow.localfs = out)
```

Comment:
We could have an R6 class that wraps around an external pointer holding a raw pointer, instead of what we usually do, i.e. an R6 class that wraps around an external pointer holding a shared pointer. 

Then we need to know if the external pointer owns it.

Reply:
@fsaintjacques you mean change the signature here https://github.com/apache/arrow/blob/master/cpp/src/arrow/dataset/file_base.h#L188

and then fix everything that breaks? Or start somewhere else?

I can try, and if I fail can make a followup jira.

Reply:
Correct, this is the place to change.

Reply:
Adding the docs really blew up the size of the diff here so I've deferred this to https://issues.apache.org/jira/browse/ARROW-7094. Will try to pick it up tomorrow.